noir/noir-repo/tooling/noirc_abi/src/lib.rs

```diff
@@ -67,6 +67,10 @@ pub enum AbiType {
     String {
         length: u64,
     },
+    FmtString {
+        length: u64,
+        item_types: Vec<AbiType>,
+    },
```

Comment:
I'm a bit leery at adding this as an `ABIType` just so that we can print error messages.

I think the cleanest solution to this is to properly typecheck constraint messages to be either string literals or format strings. We can then specialize `error_types` to hold just strings/format strings and avoid this leaking into the rest of the ABI.

Reply:
But we don't want just to have strings and format strings in assertion payloads, we do want to be able to panic with payload, same as solidity can revert with a struct error

Reply:
I've just pushed a middle ground solution: we don't have fmtstring in AbiType but we allow users to define errors with either fmtstrings or ABI-Valid types.